packages/next-server/lib/router/router.ts

```diff
@@ -73,15 +73,20 @@ export default class Router implements IRouterInterface {
 
       window.addEventListener('popstate', this.onPopState)
 
-      // Workaround for weird Firefox bug, see below links
-      // https://github.com/zeit/next.js/issues/3817
-      // https://bugzilla.mozilla.org/show_bug.cgi?id=1422334
-      // TODO: let's remove this once the Firefox bug is resolved
-      if (navigator.userAgent && navigator.userAgent.match(/firefox/i)) {
-        window.addEventListener('unload', () => {
-          if (window.location.search) window.location.replace(window.location.toString())
-        })
-      }
+      window.addEventListener('unload', () => {
+        // Workaround for weird Firefox bug, see below links
+        // https://github.com/zeit/next.js/issues/3817
+        // https://bugzilla.mozilla.org/show_bug.cgi?id=1422334
+        // TODO: let's remove this once the Firefox bug is resolved
+        if (window.location.search && navigator.userAgent &&
+          navigator.userAgent.match(/firefox/i)) {
+          window.location.replace(window.location.toString())
+        }
+        // Workaround for popstate firing on initial page load when
```

Comment:
Just so that I understand this correctly, does this mean that coming back from an external site  re-requests the page and **then** fire popstate? Because in that case it would be easy to detect using `constructor` right 🤔

Reply:
So the problem occurs when first navigating to one pages e.g. `/` then navigating with `Link` to another `/about` and then clicking an external link `google.com` and then clicking the back button. 

When you click the back button it loads the page from the server and then in some browsers like safari it fires a `popState` event with the state from before navigating to the external site.

The most simple way I've found to detect this is to tag the history state on `unload` so that we know to ignore that `popState` event.